File: asell.py
"""
This code Will Help You sell All Balance (Selected Coin)
You can change coin base on line 64-70
For More information please read API.md
"""

import requests
import hashlib
import hmac
from urllib.parse import urlencode

API_KEY = input("Please Enter Your API Key : ")
SECRET_KEY = input("Please Enter Your Secret Key : ")

# =========================
# GET WALLET BALANCE (correct endpoint)
# =========================
wallet_url = "https://rest.bit24.cash/asset/capi/v1/wallet/assets"
wallet_headers = {
    "Accept": "application/json",
    "X-BIT24-APIKEY": API_KEY
}

# Fetch all assets with positive balances, excluding IRT (fiat)
wallet_params = {
    "without_irt": "0",    # include IRT? 0 = include all, 1 = exclude IRT
    "without_zero": "1"    # only show coins with balance > 0
}
wallet_response = requests.get(
    wallet_url,
    headers=wallet_headers,
    params=wallet_params
)

if wallet_response.status_code != 200:
    print("Wallet fetch failed:")
    print(wallet_response.text)
    exit()

wallet_data = wallet_response.json()
# The data structure has "data" -> "asset" (an array)
assets = wallet_data.get("data", {}).get("asset", [])

# =========================
# FIND ADA BALANCE (available balance)
# =========================
ada_available = 0.0

for coin in assets:
    if coin["symbol"] == "ADA":
        ada_available = float(coin["available_balance"])
        break

print(f"ADA available balance: {ada_available}")

if ada_available <= 0:
    print("No ADA available to sell.")
    exit()

# =========================
# SELL ALL ADA (market sell)
# =========================
sell_url = "https://rest.bit24.cash/pro/capi/v1/orders/submit"

params = {
    "base_coin_symbol": "ADA",
    "quote_coin_symbol": "IRT",
    "type": "0",              # 0 = sell
    "category_type": "1",     # 1 = market (instant)
    "amount": str(ada_available)   # sell entire available balance
}

# Build signature: sort keys alphabetically, join with '&', HMAC-SHA256
query_string = "&".join(f"{k}={params[k]}" for k in sorted(params))
signature = hmac.new(
    SECRET_KEY.encode('utf-8'),
    query_string.encode('utf-8'),
    hashlib.sha256
).hexdigest()

params["signature"] = signature

sell_headers = {
    "Accept": "application/json",
    "Content-Type": "application/x-www-form-urlencoded",
    "X-BIT24-APIKEY": API_KEY
}

sell_response = requests.post(
    sell_url,
    headers=sell_headers,
    data=params
)

print("Sell response status:", sell_response.status_code)
print(sell_response.text)

if sell_response.status_code == 200:
    resp_json = sell_response.json()
    if resp_json.get("success"):
        print(f"Order placed successfully! Order ID: {resp_json['data']['order']['id']}")
    else:
        print(f"Order error: {resp_json.get('error')}")

File: node.js/asell.js
/**
 * This code will help you sell all balance of a selected coin
 * You can change the coin near the bottom
 *
 * For more information please read API.md
 */

const crypto = require("crypto");
const readline = require("readline");

const BASE_URL = "https://rest.bit24.cash";

// =========================
// INPUT
// =========================

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function ask(question) {
    return new Promise((resolve) => {
        rl.question(question, resolve);
    });
}

// =========================
// SIGNATURE
// =========================

function createSignature(params, secretKey) {

    const sortedKeys = Object.keys(params).sort();

    const queryString = sortedKeys
        .map((key) => `${key}=${params[key]}`)
        .join("&");

    return crypto
        .createHmac("sha256", secretKey)
        .update(queryString)
        .digest("hex");
}

// =========================
// GET WALLET BALANCES
// =========================

async function getWalletAssets(apiKey) {

    const url = new URL(
        `${BASE_URL}/asset/capi/v1/wallet/assets`
    );

    url.searchParams.append("without_irt", "0");
    url.searchParams.append("without_zero", "1");

    const response = await fetch(url, {
        method: "GET",
        headers: {
            "Accept": "application/json",
            "X-BIT24-APIKEY": apiKey
        }
    });

    const text = await response.text();

    if (!response.ok) {
        console.log("Wallet fetch failed:");
        console.log(text);

        throw new Error(`HTTP Error: ${response.status}`);
    }

    return JSON.parse(text);
}

// =========================
// FIND COIN BALANCE
// =========================

function findCoinBalance(assets, coinSymbol) {

    for (const coin of assets) {

        if (coin.symbol === coinSymbol) {
            return parseFloat(coin.available_balance);
        }
    }

    return 0;
}

// =========================
// SELL ALL BALANCE
// =========================

async function sellAllBalance(
    apiKey,
    secretKey,
    baseCoin,
    quoteCoin,
    amount
) {

    const url = `${BASE_URL}/pro/capi/v1/orders/submit`;

    const params = {
        base_coin_symbol: baseCoin,
        quote_coin_symbol: quoteCoin,

        // SELL
        type: "0",

        // MARKET
        category_type: "1",

        // Full balance
        amount: String(amount)
    };

    params.signature = createSignature(params, secretKey);

    const body = new URLSearchParams(params);

    const response = await fetch(url, {
        method: "POST",
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/x-www-form-urlencoded",
            "X-BIT24-APIKEY": apiKey
        },
        body: body.toString()
    });

    const text = await response.text();

    console.log("Sell response status:", response.status);
    console.log(text);

    if (!response.ok) {
        throw new Error(`HTTP Error: ${response.status}`);
    }

    return JSON.parse(text);
}

// =========================
// MAIN
// =========================

async function main() {

    try {

        const apiKey = await ask("Please Enter Your API Key : ");
        const secretKey = await ask("Please Enter Your Secret Key : ");

        rl.close();

        // =========================
        // CHANGE THESE
        // =========================

        const baseCoin = "ADA";
        const quoteCoin = "IRT";

        // =========================
        // FETCH WALLET
        // =========================

        console.log("Fetching wallet balances...");

        const walletData = await getWalletAssets(apiKey.trim());

        const assets = walletData?.data?.asset || [];

        // =========================
        // FIND AVAILABLE BALANCE
        // =========================

        const availableBalance = findCoinBalance(
            assets,
            baseCoin
        );

        console.log(
            `${baseCoin} available balance: ${availableBalance}`
        );

        if (availableBalance <= 0) {
            console.log(`No ${baseCoin} available to sell.`);
            return;
        }

        // =========================
        // SELL ALL
        // =========================

        console.log(
            `Selling all ${baseCoin} balance on ${quoteCoin} market...`
        );

        const result = await sellAllBalance(
            apiKey.trim(),
            secretKey.trim(),
            baseCoin,
            quoteCoin,
            availableBalance
        );

        if (result.success) {

            console.log("Order placed successfully!");

            if (result.data?.order) {

                const order = result.data.order;

                console.log(`Order ID: ${order.id}`);
                console.log(`Status: ${order.status_text}`);
                console.log(`Amount: ${order.amount}`);
                console.log(`Total: ${order.total}`);
            }

        } else {

            console.log("Order error:");
            console.log(result.error);
        }

    } catch (error) {

        console.log(`An error occurred: ${error.message}`);
    }
}

main();